ProjectBank/Infrastructure/ProjectRepository.cs

```diff
@@ -1,5 +1,157 @@
+        {
+            Description = project.Description,
+            CreationDate = DateTime.Now,
+            UpdatedDate = DateTime.Now,
+            Participants = new List<User>()
+        };
+        if (project.Tags != null)
+        {
+            entity.Tags = await SetTagsAsync(project.Tags);
+        }
+
+        _context.Projects.Add(entity);
+
+        await _context.SaveChangesAsync();
+
+        return new ProjectDTO(
+            entity.Id,
+            entity.Title,
+            entity.Status,
+            entity.UserId
+        );
+    }
+
+    public async Task DeleteProjectByIdAsync(int projectId)
+    {
+        var entity = await _context.Projects.FindAsync(projectId);
+        if (entity != null)
+        {
+            _context.Projects.Remove(entity);
+            await _context.SaveChangesAsync();
+        }
+    }
+
+    public async Task EditProjectAsync(UpdateProjectDTO project)
+    {
+        var entity = await _context.Projects.FindAsync(project.Id);
+        //Should projectDTO title be nullable? We need it for create but not for update
+        entity.Title = project.Title;
+        entity.Description = project.Description;
+        entity.UpdatedDate = DateTime.Now;
+        if (project.Tags != null)
+        {
+            entity.Tags = await SetTagsAsync(project.Tags);
+        }
+        await _context.SaveChangesAsync();
+    }
+
+    public async Task<IReadOnlyCollection<ProjectDTO>> ReadAllAsync()
+    {
+        return await _context.Projects.Select(project => new ProjectDTO(
+               project.Id,
+               project.Title,
+               project.Status,
+               project.UserId))
+               .ToListAsync();
+    }
+
+    public async Task<ProjectDTO> ReadProjectByIdAsync(int projectId)
+    {
+
+        var entity = await _context.Projects.FindAsync(projectId);
+        if (entity != null)
+        {
+            return new ProjectDTO(
+                entity.Id,
+                entity.Title,
+                entity.Status,
+                entity.UserId
+            );
+        }
+        else
+        {
+            return null;
+        }
+    }
+
+    public async Task<IReadOnlyCollection<ProjectDTO>> ReadProjectsByTagIdAsync(int tagId)
+    {
+        var tag = await _context.Tags.FindAsync(tagId);
+        return (from project in tag.Projects
+                select new ProjectDTO(
+                project.Id,
+                project.Title,
+                project.Status,
+                project.UserId)).ToList();
+    }
+
+    public async Task<IReadOnlyCollection<ProjectDTO>> ReadProjectsByUserIdAsync(int userId)
+    {
+        return await _context.Projects.Where(project => project.UserId == userId).Select(project => new ProjectDTO(
+               project.Id,
+               project.Title,
+               project.Status,
+               project.UserId))
+               .ToListAsync();
+    }
+
+    public async Task UpdateProjectStatusByIdAsync(int projectId)
```

Comment:
If it's only possible to set the status to closed shouldn't the method be called something like:
```suggestion
    public async Task SetProjectStatusToClosedByIdAsync(int projectId)
```

Reply:
we did this in the other branch :)